Fix rigid body force computation with MPI

joaander opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

<!-- Describe your changes in detail. -->
Rewrite the rigid body communication scheme so that `F_{ij} = -F_{ji}` for interactions that involve at least one constituent particle near the domain boundary.

Reviewers: Note that I documented the details of each change in the full commit log. Check that first for more details on these changes.

## Motivation and context

<!--- Why is this change required? What problem does it solve? -->
Previously, HOOMD was computing `F_{ij}` as a function of `\vec{r}_i(t-1)` and `\vec{r}_j(t)` for interactions involving some constituent particles near the domain boundaries. There were many errors in the previous communication scheme.

<!-- Replace ??? with the issue number that this pull request resolves. -->
Resolves #1468

## How has this been tested?

<!--- Please describe in detail how you tested your changes. -->
See the minimal test case in #1468.

I am running longer momentum/energy conservation tests now.

<!--- Please build the sphinx documentation and check that any changes to
      documentation display properly. -->

## Change log

<!-- Propose a change log entry. -->
```
Fixed:

* Compute pairwise forces that add to zero for rigid body constituent particles near MPI domain boundaries.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

joaander committed:
Update rigid bodies both before and after communication.

This ensures that constituent particles are migrated to the proper
domain before forces are calculated. Without the call before, the
rigid body centers are moved but the constituents left at the previous
position. In this way, a constituent that should migrate to a new rank
might be left behind - outside the ghost particle window for a neighbor
rank. This leads to i->j force calculations without the corresponding
j->i calculation on a neighbor rank.

By updating bodies before communicating, the constituent particles
that move outside the local domain are communicated correctly. The
update bodies call after communication is required to ensure that
constituents are updated properly on ranks where both the constituent
AND the central particle are ghosts.

joaander committed:
Correct the computation of the body ghost width.

Constituent particles need no extra ghost width. Only those within
the normal r_cut + r_buff range of a boundary need to be communicated.

Only the rigid body central particles *may* need an extra ghost
width. Specifically, the worse case scenario is when a constituent
ghost is exactly r_ghost_width away from the boundary. To compute
the correct position, the central particle must also be a ghost,
so the minimum ghost width for a central particle is
max[d_i + r_ghost_i] over all particles in the body. This
sets a minimum. If the ghost width for the body central particle is
already greater than this, it need not be increased.

joaander committed:
Run pre-commit.

joaander committed:
Compute maximum ghost width appropriately.

Evaluate each per type ghost with separately and return the maximum
of those.

joaander committed:
Document the rigid body communication scheme.

joaander committed:
Remove neighborlist error based on rigid body diameters.

This error was incorrectly determined. The minimum image convention
is applied only between individual particles, not bodies. If the user
chooses to make a rigid body as large as the box, the simulation may
be ill-stated but should run. There isn't much HOOMD can do to protect
users from the egregious error of defining a rigid body larger than the
simulation box.

jekw90 commented:
Thank you for fixing rigid body force computation with MPI. This is really nice! I will do my oral defense soon, and I have been busy writing thesis & papers. So, I think it would be really great if other peeps review this pull-request. Thank you.

joaander commented:
@jekw90 Thanks for letting me know quickly, I have assigned another reviewer. @rayasare you are up.

joaander committed:
Add body ghost width for constituents.

In previous commits on this branch, I incorrectly assumed that the
body ghost width for constituents should be 0. My test case involved
bodies where the r_cut for constituents was large. The body net force
computation requires that the rank hosting the central particle must
be able to access all constituents as ghosts so that it can sum the
net force and torque. Adjust the body ghost width computation to
account for this.

tommy-waltmann commented:
Are the longer momentum/energy conservation tests in `hoomd-validation`?

joaander commented:
> Are the longer momentum/energy conservation tests in `hoomd-validation`?

Yes: https://github.com/glotzerlab/hoomd-validation/tree/rigid-bodies

joaander commented:
While it is much improved compared to the first run (https://github.com/glotzerlab/hoomd-blue/pull/1424#issuecomment-1398592517), there is still some energy and momentum drift:
![image](https://user-images.githubusercontent.com/2197568/214571454-f4a2a1d9-af56-493d-a15a-8224f42e2d90.png)

I don't have ideas where the new problem could be at this time. My minimal test no longer shows any detectable differences between serial and MPI CPU execution.

There must be some very small, but systematic error in the net force on the system to give it a constantly increasing momentum to cause this drift. At this point, we need to start over with the debugging process and check serial CPU vs GPU and MPI GPU, with and without rotational degrees of freedom, logging the net force over time to see if there is the systematic error I suspect and where it might be coming from. I will open a new issue and we can fix it later in another bugfix PR. Let's not let the remaining issue prevent us from merging this fix that covers the egregious errors.

tommy-waltmann commented on hoomd/md/IntegratorTwoStep.cc:
```diff
@@ -75,21 +75,25 @@ void IntegratorTwoStep::update(uint64_t timestep)
     else
 #endif
-        {
+
+        // Update rigid body constituent particles after communicating (or once in serial) to ensure
+        // that all ghost constituent particle positions are set in accordance with any just
+        // communicated ghost and/or migrated rigid body centers.
         updateRigidBodies(timestep + 1);
```
Shouldn't be there be a set of open and closed brackets here after the `else`? This is in one other place in this PR too. I'm surprised a linter didn't find this.

joaander replied:
Yes, will fix.

tommy-waltmann reviewed (changes requested):
One small code formatting thing before we merge this.

joaander commented:
I found that my minimal test does detect the issue - but only when run on 4 or more ranks. The issue is not present on two ranks:
```
$ mpirun -n 4 python test_nve_rigid.py
notice(2): Using domain decomposition: n_x = 1 n_y = 2 n_z = 2.
box.Lz/2 = 11.696070671081543
E_0=525.0778866847149
p_0=(-1.6342482922482304e-13, -4.618527782440651e-14, -1.6697754290362354e-13)
net lj force: [-0.00011244  0.00080547 -0.00074825]
net rigid force: [-0.00011244  0.00080547 -0.00074825]
```

```
$ mpirun -n 2 python test_nve_rigid.py
notice(2): Using domain decomposition: n_x = 1 n_y = 1 n_z = 2.
box.Lz/2 = 11.696070671081543
E_0=525.0778866847152
p_0=(2.1316282072803006e-14, 3.375077994860476e-14, -1.4566126083082054e-13)
net lj force: [-4.16389145e-13  5.41788836e-14 -1.11632925e-13]
net rigid force: [-8.88178420e-15  5.10702591e-15 -2.62012634e-14]
```

Will delay merging this until I can find a fix.

joaander committed:
Add brackets around updateRigidBodies serial condition.

The compute callback calls updateRigidBodies again when needed in
MPI domain decompostion runs. Add comments describing this process.

joaander committed:
Always issue an error for missing central particles.

In the new scheme, even ghost constituent particles need to access
a ghost central particle.

joaander committed:
Remove unused getBodyDiameter.

joaander committed:
Correct additional corner cases.

In the previous commits on this branch, I assumed that the rigid body
did not need to be complete. This assumption was incorrect, as the way
that ForceComposite indexes particles from the molecule in reference to
the body assumes that the body is complete. Expand the constituent
ghost width to include this.

With the additional ghost width, there are now many extra constituent
ghost particles that cannot access their central particle. However, they
do not need to because they are outside the pair force interaction range
of the domain boundary. Therefore, we need to allow for non-local
central particles. By construction of the body ghost widths, any non-local
central particle should not be needed. However, there is no straightforward
way to check for this error condition as updateCompositeParticles does not
have knowledge of the interaction ghost widths to verify that every
constituent ghost within the interaction width has a central particle.

joaander commented:
Fix pushed, tested with the minimal test script on 2, 4, 8 CPUs *and* 2,4,8 GPUs. Re-running the long conservation tests now.

joaander commented:
Success!
![image](https://user-images.githubusercontent.com/2197568/214833786-8330fdb8-5a11-47d2-a2fd-a5b0c8ec2a5d.png)
